src/files/pyfingerprint/pyfingerprint.py

```diff
@@ -673,8 +677,86 @@ def readImage(self):
         else:
             raise Exception('Unknown error '+ hex(receivedPacketPayload[0]))
 
-    ## TODO:
-    ## Implementation of uploadImage()
+    def uploadImage(self, imageLocation):
+    """
+    Upload the image of a finger from host computer.
+
+    @param string imageLocation
+    @return void
+    """
+    locationDirectory = os.path.dirname(imageLocation)
+    if not os.access(locationDirectory, os.R_OK):
+        raise ValueError('The given location directory "' + locationDirectory + '" is not readable!')
+
+    if not os.path.exists(imageLocation):
+        raise ValueError('File "' + imageLocation + '" does not exist!')
+
+    maxPacketSize = self.getMaxPacketSize()
+
+    print("Sending first packet to upload image...") # debug
+    packetPayload = (
+        FINGERPRINT_UPLOADIMAGE,
+    )
+
+    self.__writePacket(FINGERPRINT_COMMANDPACKET, packetPayload)
+
+    ## Get first reply packet
+    receivedPacket = self.__readPacket()
+
+    receivedPacketType = receivedPacket[0]
+    receivedPacketPayload = receivedPacket[1]
+
+    if receivedPacketType != FINGERPRINT_ACKPACKET:
+       raise Exception('The received packet is no ack packet!')
+    # DEBUG: The sensor will send follow-up packets
+    if receivedPacketPayload[0] == FINGERPRINT_OK:
+        pass
+    elif receivedPacketPayload[0] == FINGERPRINT_ERROR_COMMUNICATION:
+        raise Exception('Communication error')
+    elif receivedPacketPayload[0] == FINGERPRINT_ERROR_UPLOADIMAGE:
+        raise Exception('Could not upload image')
+    else:
+        raise Exception('Unknown error ' + hex(receivedPacketPayload[0]))
+
+    # Initialize image library
+    imageToUpload = Image.open(imageLocation)
+    imageArray = numpy.array(imageToUpload)
+
+    maxRowLength = max(map(lambda x: len(x), imageArray))
+
+    print("will start sending data packets")
```

Comment:
Maybe convert some of these debug prints to comments to avoid spamming the user when using this method. The information as a comment is helpful, but maybe should not be printed if it works.